### PR TITLE
Send email to project users on sample registration

### PR DESCRIPTION
**Pull request by wow-such-code**
* adds predefined email message that explains sample registration happened
* creates file from sample batch metadata to attach to message
* adds possibility to send attachments with emails
* sends email to every active user having access to the project upon successful sample registration

### CODE REVIEW

File: domain-concept/src/main/java/life/qbic/domain/concepts/communication/CommunicationService.java [hunk 1]
```diff
@@ -11,4 +11,6 @@ public interface CommunicationService {
 
   void send(Subject subject, Recipient recipient, Content content) throws CommunicationException;
 
+  void send(Subject subject, Recipient recipient, Content content, String attachmentContent,
```
Comment (sven1103) [inline domain-concept/src/main/java/life/qbic/domain/concepts/communication/CommunicationService.java:14]: I suggest to introduce a record 'attachment' to keep the parameter list short.

File: projectmanagement-domain/src/main/java/life/qbic/projectmanagement/application/authorization/acl/ProjectAccessService.java [hunk 2]
```diff
@@ -21,6 +21,14 @@ public interface ProjectAccessService {
    */
   List<String> listUsers(ProjectId projectId);
 
+  /**
+   * Lists all active users which have a permission within the specific project
+   *
+   * @param projectId the identifier of the project
+   * @return a list of user ids which are associated with the project
+   */
+  List<String> listActiveUsers(ProjectId projectId);
```
Comment (sven1103) [inline projectmanagement-domain/src/main/java/life/qbic/projectmanagement/application/authorization/acl/ProjectAccessService.java:30]: nice :+1:

File: projectmanagement-domain/src/main/java/life/qbic/projectmanagement/application/policy/directive/InformUsersAboutBatchRegistration.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+  @Job(name = "Notify users about batch registration")
+  /**
+   * Sends an email with attached spreadsheet of newly registered samples to the users of a project.
+   * The email contains an explanation about sample identifiers and the spreadsheet the known metadata of the samples
+   * @param project - the project the sample batch was added to
+   * @param sampleBatch - the samples that were registered
+   */
+  public void notifyUsersAboutSamples(Project project, Collection<Sample> sampleBatch) {
+    List<RecipientDTO> recipientsWithAccess = getUsersWithAccess(project.getId());
```
Comment (sven1103) [inline projectmanagement-domain/src/main/java/life/qbic/projectmanagement/application/policy/directive/InformUsersAboutBatchRegistration.java:70]: ah sorry, my mistake. The method promises to return users, not recipients. So you might need a static wrapper method like ``userToRecipient(UserInfo user)``

File: projectmanagement-domain/src/main/java/life/qbic/projectmanagement/application/policy/directive/InformUsersAboutBatchRegistration.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  private void notifyRecipient(RecipientDTO recipient, Project project, String attachmentContent) {
+    String subject = "New samples added to project";
+    String projectUri = project.getId().toString();
+    String projectTitle = project.getProjectIntent().projectTitle().title();
+
+    var message = Messages.samplesAddedToProject(recipient.getFullName(), projectTitle, projectUri);
+
+    communicationService.send(new Subject(subject),
+        new Recipient(recipient.getEmailAddress(), recipient.getFullName()), new Content(message),
+        attachmentContent, "sample_sheet.tsv");
```
Comment (sven1103) [inline projectmanagement-domain/src/main/java/life/qbic/projectmanagement/application/policy/directive/InformUsersAboutBatchRegistration.java:115]: that is QBiC legacy terminology, I suggest sth like ``batch-<name>-samples``

File: projectmanagement-domain/src/main/java/life/qbic/projectmanagement/domain/project/sample/event/SampleBatchRegistered.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import java.io.Serial;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.Objects;
+import life.qbic.domain.concepts.DomainEvent;
+import life.qbic.projectmanagement.domain.project.Project;
+import life.qbic.projectmanagement.domain.project.sample.BatchId;
+import life.qbic.projectmanagement.domain.project.sample.Sample;
+
+/**
```
Comment (sven1103) [inline projectmanagement-domain/src/main/java/life/qbic/projectmanagement/domain/project/sample/event/SampleBatchRegistered.java:12]: We already have a BatchRegistered Domain Event that must be reused here

**Review by sven1103 — CHANGES_REQUESTED** [review]
Thank you @wow-such-code for your contribution.

I have some concernts, but I think they are easily resolvable. Let me know if sth is unclear.

File: domain-concept/src/main/java/life/qbic/domain/concepts/communication/CommunicationService.java [hunk 6]
```diff
@@ -11,4 +11,5 @@ public interface CommunicationService {
 
   void send(Subject subject, Recipient recipient, Content content) throws CommunicationException;
 
+  void send(Subject subject, Recipient recipient, Content content, Attachment attachment) throws CommunicationException;
```
Comment (sven1103) [inline domain-concept/src/main/java/life/qbic/domain/concepts/communication/CommunicationService.java:14]: Maybe some JDs :D

File: projectmanagement-domain/src/main/java/life/qbic/projectmanagement/domain/project/service/BatchDomainService.java [hunk 7]
```diff
@@ -35,29 +38,25 @@ public BatchDomainService(BatchRepository batchRepository) {
...
+    if (result.isError()) {
+      return Result.fromError(ResponseCode.BATCH_CREATION_FAILED);
+    } else {
+      dispatchSuccessfulSampleBatchRegistration(label, batch.batchId(), project);
+    }
+    return Result.fromValue(result.getValue().batchId());
   }
 
 
+  private void dispatchSuccessfulSampleBatchRegistration(String name, BatchId id, Project project) {
```
Comment (sven1103) [inline projectmanagement-domain/src/main/java/life/qbic/projectmanagement/domain/project/service/BatchDomainService.java:58]: NIT:

```suggestion
  private void dispatchRegistration(String name, BatchId id, Project project) {
```

**Review by sven1103 — CHANGES_REQUESTED** [review]
One last request: domain events should be slim with simple data types to ensure serialization and not passing too much information around.

Instead of the project aggregate, you can pass the project id.

Second: why is the project id needed?

**Review by Steffengreiner — APPROVED** [review]
Nice Spot with the distinct users! LGTM 🚀

**Pull request merged**